src/amo/components/Categories.js

```diff
@@ -1,29 +1,53 @@
 import React, { PropTypes } from 'react';
+import { connect } from 'react-redux';
 import { compose } from 'redux';
+import { showLoading } from 'react-redux-loading-bar';
 
 import Link from 'amo/components/Link';
-import { visibleAddonType } from 'core/utils';
+import { categoriesFetch } from 'core/actions/categories';
+import { apiAddonType, visibleAddonType } from 'core/utils';
 import translate from 'core/i18n/translate';
+import LoadingText from 'ui/components/LoadingText';
 
 import './Categories.scss';
 
 
 export class CategoriesBase extends React.Component {
   static propTypes = {
     addonType: PropTypes.string.isRequired,
+    dispatch: PropTypes.func,
     categories: PropTypes.object.isRequired,
+    clientApp: PropTypes.string.isRequired,
     error: PropTypes.bool,
     loading: PropTypes.bool.isRequired,
     i18n: PropTypes.object.isRequired,
   }
 
+  componentWillMount() {
+    const { addonType, categories, clientApp, dispatch } = this.props;
+    if (!Object.values(categories).length) {
+      dispatch(showLoading());
+      dispatch(categoriesFetch({ addonType, clientApp }));
+    }
+  }
+
   render() {
     const { addonType, categories, error, loading, i18n } = this.props;
 
     if (loading) {
       return (
         <div className="Categories">
-          <p>{i18n.gettext('Loading...')}</p>
+          <span className="visually-hidden">{i18n.gettext('Loading…')}</span>
+          <ul className="Categories-list"
```

Comment:
I think some of this could be refactored so the categories were just an array of `<LoadingText />` elements and if `loading === true` they are displayed in `<span>` instead of `<a>` tags. I can refactor this to duplicate less markup I think.